### PR TITLE
JDK-8320405: [Windows Server 2016] java/awt/image/MultiResolutionImage/MultiResolutionImageObserverTest.java shows issues in awt_Win32GraphicsDevice.cpp

### PR DESCRIPTION
**Pull request by MBaesken**
When running with fastdebug binaries we run intermittent into the issue below in
jtreg test java/awt/image/MultiResolutionImage/MultiResolutionImageObserverTest.java .
Seems we miss checking of successful HBITMAP creation before calling GetDIBits.

  HDC hBMDC = this->GetDC();
  HBITMAP hBM = ::CreateCompatibleBitmap(hBMDC, 1, 1);
  VERIFY(::GetDIBits(hBMDC, hBM, 0, 1, NULL, gpBitmapInfo, DIB_RGB_COLORS));

in awt_Win32GraphicsDevice.cpp . Thats why the releast of hBMDC / hBM fails as well at the end of the function causing the seond and third warning.

Sat Nov 18 17:29:23 CET 2023

*********************
AWT Assertion Failure
*********************
::GetDIBits(hBMDC, hBM, 0, 1, 0, gpBitmapInfo, 0)
File 'e:\openjdk\openjdk-21u-windows_x86_64-dbg\jdk\src\java.desktop\windows\native\libawt\windows\awt_Win32GraphicsDevice.cpp', at line 184
GetLastError() is 57 : The parameter is incorrect.

Do you want to break into the debugger?
*********************
*********************
AWT Assertion Failure
*********************
::DeleteObject(hBM)
File 'e:\openjdk\openjdk-21u-windows_x86_64-dbg\jdk\src\java.desktop\windows\native\libawt\windows\awt_Win32GraphicsDevice.cpp', at line 297
GetLastError() is 57 : The parameter is incorrect.

So it seems, we should have some additional checks/tracing.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8320405](https://bugs.openjdk.org/browse/JDK-8320405): [Windows Server 2016] java/awt/image/MultiResolutionImage/MultiResolutionImageObserverTest.java shows issues in awt_Win32GraphicsDevice.cpp (**Bug** - P4) ⚠️ Issue is not open.

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17197/head:pull/17197` \
`$ git checkout pull/17197`

Update a local copy of the PR: \
`$ git checkout pull/17197` \
`$ git pull https://git.openjdk.org/jdk.git pull/17197/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17197`

View PR using the GUI difftool: \
`$ git pr show -t 17197`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17197.diff">https://git.openjdk.org/jdk/pull/17197.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17197#issuecomment-1871008622)

### CODE REVIEW

File: src/java.desktop/windows/native/libawt/windows/awt_Win32GraphicsDevice.cpp [hunk 1]
```diff
@@ -137,7 +137,13 @@ HDC AwtWin32GraphicsDevice::MakeDCFromMonitor(HMONITOR hmMonitor) {
             HDC hDC = CreateDC(mieInfo.szDevice, NULL, NULL, NULL);
             if (NULL != hDC) {
                 retCode = hDC;
+            } else {
+                J2dTraceLn1(J2D_TRACE_WARNING,
+                       "CreateDC failed, so we cannot store the DC for later usage, mieInfo.szDevice is %S",
```
Comment (aivanov-jdk) [inline src/java.desktop/windows/native/libawt/windows/awt_Win32GraphicsDevice.cpp:142]: Is `%S` a valid format string? It should in lower case: `%s`.

It seems, “store the DC for later usage” is misleading, it isn't stored… at least for a long period of time.

File: src/java.desktop/windows/native/libawt/windows/awt_Win32GraphicsDevice.cpp [hunk 2]
```diff
@@ -179,7 +185,13 @@ void AwtWin32GraphicsDevice::Initialize()
     }
     gpBitmapInfo->bmiHeader.biBitCount = 0;
     HDC hBMDC = this->GetDC();
+    if (hBMDC == NULL) {
+        J2dTraceLn(J2D_TRACE_WARNING, "AwtWin32GraphicsDevice::Initialize GetDC failed");
+    }
     HBITMAP hBM = ::CreateCompatibleBitmap(hBMDC, 1, 1);
+    if (hBM == NULL) {
+        J2dTraceLn(J2D_TRACE_WARNING, "AwtWin32GraphicsDevice::Initialize CreateCompatibleBitmap failed");
+    }
```
Comment (aivanov-jdk) [inline src/java.desktop/windows/native/libawt/windows/awt_Win32GraphicsDevice.cpp:194]: Does it make sense to call `::GetLastError()` and trace its value?

It seems that we cannot continue if `hBMDC` or `hBM` are null. Yet it seems unexpected.

File: src/java.desktop/windows/native/libawt/windows/awt_Win32GraphicsDevice.cpp [hunk 3]
```diff
@@ -293,8 +305,8 @@ void AwtWin32GraphicsDevice::Initialize()
         }
         palette->UpdateLogical();
     }
-    VERIFY(::DeleteObject(hBM));
-    VERIFY(::DeleteDC(hBMDC));
+    if (hBM != NULL) VERIFY(::DeleteObject(hBM));
+    if (hBMDC != NULL) VERIFY(::DeleteDC(hBMDC));
```
Comment (aivanov-jdk) [inline src/java.desktop/windows/native/libawt/windows/awt_Win32GraphicsDevice.cpp:309]: This should rather be expanded `if` statements with braces.

**Review by aivanov-jdk — COMMENTED** [review]
Is the purpose of this review to just add traces?

The errors seems to come from *headless* environment. A better fix would be to add `@key headful` to the `MultiResolutionImageObserverTest.java` test, even though the test does not display UI.

File: src/java.desktop/windows/native/libawt/windows/awt_Win32GraphicsDevice.cpp (hunk 2 above)
Comment (aivanov-jdk) [reply src/java.desktop/windows/native/libawt/windows/awt_Win32GraphicsDevice.cpp:194]: > It seems that we cannot continue if `hBMDC` or `hBM` are null. Yet it seems unexpected.

Should we rather modify the code to back out and not to call APIs with NULL handles?

Comment (mrserb) [reply src/java.desktop/windows/native/libawt/windows/awt_Win32GraphicsDevice.cpp:194]: I think we should fallback to something. The bug itself looks similar to this: https://bugs.openjdk.org/browse/JDK-8185862

**Comment by MBaesken** [comment]
> Is the purpose of this review to just add traces?
> 
> The errors seems to come from _headless_ environment. A better fix would be to add `@key headful` to the `MultiResolutionImageObserverTest.java` test, even though the test does not display UI.

Hi Alexey, I am not against adding the key to the test, probably this makes sense.
But still I think the coding itself should work better also in headless mode.

File: src/java.desktop/windows/native/libawt/windows/awt_Win32GraphicsDevice.cpp (hunk 1 above)
Comment (MBaesken) [reply src/java.desktop/windows/native/libawt/windows/awt_Win32GraphicsDevice.cpp:142]: https://learn.microsoft.com/en-us/cpp/c-runtime-library/format-specification-syntax-printf-and-wprintf-functions?view=msvc-170
The %S is for wide chars, do you think we need the other one (%s) ?

File: src/java.desktop/windows/native/libawt/windows/awt_Win32GraphicsDevice.cpp (hunk 2 above)
Comment (aivanov-jdk) [reply src/java.desktop/windows/native/libawt/windows/awt_Win32GraphicsDevice.cpp:194]: > I think we should fallback to something. The bug itself looks similar to this: https://bugs.openjdk.org/browse/JDK-8185862

[JDK-8185862](https://bugs.openjdk.org/browse/JDK-8185862) seems to be the exact duplicate of this one.

It references [JDK-8269529](https://bugs.openjdk.org/browse/JDK-8269529) which implemented [a workaround](https://github.com/openjdk/jdk/commit/d042029509a8cbdb723f78e2cfee4e2885775814#diff-fba4a205d0e60634be77d7a7f16a582277f7461d92a9c70cfb96d4a87068d0f1R36):

https://github.com/openjdk/jdk/blob/d042029509a8cbdb723f78e2cfee4e2885775814/test/jdk/javax/swing/reliability/HangDuringStaticInitialization.java#L36

[The same workaround](https://github.com/openjdk/jdk/commit/5c083e8560ce9cc78615e3149a558206724cff53#diff-3b5e63f29c2a52feda881e4115e5d09d1b8aecd4698f1d72981de94caa3a46d0R50) was implemented for [JDK-8266129](https://bugs.openjdk.org/browse/JDK-8266129):

https://github.com/openjdk/jdk/blob/5c083e8560ce9cc78615e3149a558206724cff53/test/jdk/tools/jpackage/windows/WinInstallerIconTest.java#L50

There's @prrace's comment [in the JBS](https://bugs.openjdk.org/browse/JDK-8266129?focusedId=14416854&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14416854): <q cite="https://bugs.openjdk.org/browse/JDK-8266129?focusedId=14416854&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14416854">There's a bunch of AWT implementation missing to properly work headless. Random tests run into it. The tests should not have to deal with it. Instead it requires an unfunded effort to rework headless mode support in AWT and 2D.</q>

So, we can add `@requires !vm.debug` to the `MultiResolutionImageObserverTest.java` test or, better, handle the situation gracefully.

**Comment by aivanov-jdk** [comment]
Hi Matthias,

> But still I think the coding itself should work better also in headless mode.

Then it should not only trace but fallback gracefully without calling APIs with the invalid parameters as discussed [in another thread](https://github.com/openjdk/jdk/pull/17197#discussion_r1439778065).

The traces won't fix anything. According to the provided logs `hBMDC` is `NULL`, therefore `hBM` also becomes `NULL`, thus `::GetDIBits` also fails.

There's already a workaround for [JDK-4684966](https://bugs.openjdk.org/browse/JDK-4684966):

https://github.com/openjdk/jdk/blob/fb90af881badf143163c7d0b9961152c2a12cd84/src/java.desktop/windows/native/libawt/windows/awt_Win32GraphicsDevice.cpp#L195-L199

It should be extended to include the possible failure in the preceding call to `::GetDIBits`.

File: src/java.desktop/windows/native/libawt/windows/awt_Win32GraphicsDevice.cpp (hunk 1 above)
Comment (aivanov-jdk) [reply src/java.desktop/windows/native/libawt/windows/awt_Win32GraphicsDevice.cpp:142]: > https://learn.microsoft.com/en-us/cpp/c-runtime-library/format-specification-syntax-printf-and-wprintf-functions?view=msvc-170 The %S is for wide chars, do you think we need the other one (%s) ?

You're right, it should be wide-char.

**Comment by MBaesken** [comment]
> It should be extended to include the possible failure in the preceding call to ::GetDIBits.

I handle now also a failure of the first GetDIBits call; and also added braces at the if's mentioned.

**Comment by aivanov-jdk** [comment]
> > It should be extended to include the possible failure in the preceding call to ::GetDIBits.
> 
> I handle now also a failure of the first GetDIBits call; and also added braces at the if's mentioned.

Sorry for my delayed reply. It's not what I expected.

What I meant is that we shouldn't even call `::CreateCompatibleBitmap` if `hBMDC == NULL`. Then we shouldn't call `::GetDIBits` if `hBM == NULL`. Thus in this case, the _fallback_ code path should be taken.

Perhaps, the code could be restructured… Yet I don't fully understand what's done here.

**Comment by RealCLanger** [comment]
I have closed [JDK-8320405](https://bugs.openjdk.org/browse/JDK-8320405) as it does not match the synopsis of the issue addressed here. [JDK-8185862](https://bugs.openjdk.org/browse/JDK-8185862) is a better fit. I suggest to also close this PR and reboot on [JDK-8185862](https://bugs.openjdk.org/browse/JDK-8185862).

Technically, I agree, if `hBMDC` is NULL, we should return right away. However, the question is, how critical such a situation is. I think we need an understanding about the environment in which that can happen and whether we can accept it without assertions.

E.g. what it means when mieInfo.szDevice=`WinDisc` in a call to `HDC hDC = CreateDC(mieInfo.szDevice, NULL, NULL, NULL); ` which obviously yields NULL for hDC.

**Comment by prrace** [comment]
There are multiple levels of problem here
At the highest level there's an entire code path we should not have entered if this is headless, but
on windows the AWT implementation is not doing a good job of this. I don't know off-hand exactly
what a fix for it would look like but changing this code without looking at the bigger picture seems premature

If we handle that, then we would not be here and would not have an issue in your environment and moreover
since we would not see the failure, we would not have an assert to deal with. 
It is then a separate issue unrelated to the debug build that we don't check for failure but I really doubt
we'd ever see a failure here on a headful config.

I do not agree with adding the headful keyword because this is a common code path and it essentially means
we'd keep adding it to a load of headful tests because of this AWT problem.

Our answer to this to date has been do not run AWT tests on debug builds.
Arguably this has "worked" as much by luck as anything else since clearly the call can fail on product builds too, but it has never caused any problems that I have heard of. That doesn't make it satisfactory, but it also doesn't make  a point fix the right fix.

**Comment by RealCLanger** [comment]
I agree, maybe in that kind of headless environment it could make sense to not initialize screens. I'll play with it and hopefully come back with something on [JDK-8185862](https://bugs.openjdk.org/browse/JDK-8185862). At least it seems we have such an environment in our configuration, so I can test there. Let's close this PR.

**Pull request closed without merging**